src/core/DY.Core.Trace.Base.fst

```diff
@@ -13,20 +13,108 @@ type trace = trace_ label
 
 /// The length of a trace.
 
-val length: #label_t:Type -> trace_ label_t -> nat
-let rec length tr =
+val trace_length: #label_t:Type -> trace_ label_t -> nat
+let rec trace_length tr =
   match tr with
   | Nil -> 0
-  | Snoc init last -> length init + 1
+  | Snoc init last -> trace_length init + 1
+
+
+(*** Hiding constructors ***)
+
+val empty_trace : #label_t:Type -> trace_ label_t 
+let empty_trace = Nil
+
+val is_empty: #label_t:Type -> trace_ label_t -> bool
+let is_empty = Nil?
+
+val is_not_empty: #label_t:Type -> trace_ label_t -> bool
+let is_not_empty = Snoc?
+
+val init : #label_t:Type -> tr:trace_ label_t{is_not_empty tr} -> trace_ label_t
+let init tr =
+  let Snoc init _ = tr in
+  init
+
+val last: #label_t:Type -> tr:trace_ label_t{is_not_empty tr} -> trace_entry_ label_t
+let last tr =
+  let Snoc _ en = tr in
+  en
+
+// to be used infix: tr `append_entry` entry
+val append_entry: #label_t:Type -> trace_ label_t -> trace_entry_ label_t -> trace_ label_t
+let append_entry tr en = Snoc tr en
+
+
+(*** To/From List Conversion ***)
+
+/// generate a trace from a list,
+/// where the first element of the list,
+/// is the last element of the trace
+/// (see the example below)
+val trace_from_rev_list: #label_t:Type -> list (trace_entry_ label_t) -> trace_ label_t
+let rec trace_from_rev_list entries =
+  match entries with
+  | [] -> Nil
+  | hd::tl -> Snoc (trace_from_rev_list tl) hd
+
+open FStar.List.Tot.Base
+
+/// generate a trace from a list,
+/// where the first element of the list,
+/// is the first element of the trace
+/// (see the example below)
+val trace_from_list: #label_t:Type -> list (trace_entry_ label_t) -> trace_ label_t
```

Comment:
There is also the question about the right names.
`from_list` could be confusing, since lists are usually read left to right, but the most recent entry on the trace will be the last list entry.
What do people think?

Reply:
In my mind the trace has old entries on the left and new entries on the right (because in western culture, time goes from left to right) so I think `trace_from_list` makes sense. I feel the behavior of `trace_from_rev_list` is less intuitive and should be considered as an auxiliary function.

Reply:
I agree, that from looking at the list and the trace, the non-rev version is more intuitive.
On the other hand, if you do `head (trace_to_list tr)` you may be surprised to not get the most recent trace entry.
We can keep the names for now, and change them later, if needed?

Reply:
This doesn't seem shocking to me, I think the current name and behavior is good.

Reply:
This also seems fine to me --- I don't expect to use `trace_to_list` all that much outside of the debug files or some tests for internal things, so since it's not very user-facing, I'm not too worried about how intuitive it is for anything other than printing. We can of course always change that later if needed.